32/RLN-SPEC

b-d1 opened the pull request:
# Proposal for an RLN RFC

b-d1 committed:
RLN SPEC RFC added

b-d1 committed:
fix: NRLN removed, equation fixes

D4nte commented:
Can you please use [Semantic Line Break](https://sembr.org/) as per our [conventions?](https://rfc.vac.dev/spec/1/#conventions)

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,365 @@
+
+# Motivation
+
+RLN (Rate limiting nullfier) is a construct based on zero-knowledge proofs that enables spam prevention mechanism for decentralized, anonymous environments. In anonymous environments, the identity of the entities is unknown, and thus it is very hard to prevent spam attacks. Many applications would benefit from anonimity, but potential for spam could worsen their UX by large extent. RLN allows for ...
```
```suggestion
RLN (Rate limiting nullfier) is a construct based on zero-knowledge proofs that enables spam prevention mechanism for decentralized, anonymous environments. In anonymous environments, the identity of the entities is unknown, and thus it is very hard to prevent spam attacks. Many applications would benefit from anonymity, but the potential for spam could worsen their UX to a large extent. RLN allows for anonymous signaling while enabling spam prevention on protocol level.
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,365 @@
+
+# Motivation
+
+RLN (Rate limiting nullfier) is a construct based on zero-knowledge proofs that enables spam prevention mechanism for decentralized, anonymous environments. In anonymous environments, the identity of the entities is unknown, and thus it is very hard to prevent spam attacks. Many applications would benefit from anonimity, but potential for spam could worsen their UX by large extent. RLN allows for ...
```
I suggest adding the definition of spam.

staheri14 replied:
Also, I think it would be good to introduce RLN as a  rate-limiting tool and mention spam protection as one of its applications e.g., it can also be used in e-voting.

b-d1 replied:
I agree, I would add those definitions and change the section accordingly.

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
feat: content updates & semantic breaks

b-d1 commented:
I've added semantic line breaks, and updated the content a bit (motivation area) to reflect the comments above. I'll update the contributors section once everything else is approved.

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+---
+slug: 32
+title: 32/RLN-SPEC
```
I'd just call this "RLN", "SPEC" seems redundant.

b-d1 replied:
Agreed

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+You can think of MetaMask for ZK-Proofs. 
+It uses `zk-kit` under the hood.
+
+
```
Don't forget the Copyright section :) Like this https://rfc.vac.dev/spec/10/#copyright

b-d1 replied:
Will be added

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+
+## ZK Circuits specification
+
+Anonymous signaling with spam detection is enabled by proving that the user is part of a group which has high bariers for entry (form of stake) and 
```
```suggestion
Anonymous signaling with spam detection is enabled by proving that the user is part of a group which has high barriers to entry (form of stake) and 
```

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+enabling secret reveal if more signals are produced than the spam threshold (system parameter) per epoch.
+The membership part is implemented using membership [merkle trees](https://en.wikipedia.org/wiki/Merkle_tree) and merkle proofs, 
+while the secret reveal part is enabled by using the Shamir's Secret Sharing scheme. 
+Esentially the protocol requires the users to generate zero-knowledge proof to be able to send signals and participate in the application. 
```
```suggestion
Essentially the protocol requires the users to generate zero-knowledge proof to be able to send signals and participate in the application. 
```

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+The zero knowledge proof proves that the user is member of a group, 
+but also enforces the user to share part of their secret for each signal in an epoch.
+The epoch is an external nullifier which is usually represented by timestamp or a time interval. 
+It can also be tought as voting booth in voting applications.
```
```suggestion
It can also be thought of as a voting booth in voting applications.
```

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+Esentially the protocol requires the users to generate zero-knowledge proof to be able to send signals and participate in the application. 
+The zero knowledge proof proves that the user is member of a group, 
+but also enforces the user to share part of their secret for each signal in an epoch.
+The epoch is an external nullifier which is usually represented by timestamp or a time interval. 
```
```suggestion
The epoch is an external nullifier which is usually represented by timestamp or a time interval. 
```

Might be a good idea to define terms like *external nullifier*, *epoch* etc in a separate section for clarity. E.g. see https://rfc.vac.dev/spec/2/#definitions or https://rfc.vac.dev/spec/13/#terminology

b-d1 replied:
yes I agree with this

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+
+### Hash function
+
+Canonical [Poseidon hash implementation](https://eprint.iacr.org/2019/458.pdf) is used, 
```
Perhaps we can write here what the canonical parameters used are?

Also, it'd be neat if we could have some basic section for security considerations / cryptographic primitives. E.g. see

https://rfc.vac.dev/spec/26/#cryptographic-primitives https://rfc.vac.dev/spec/10/#primary-adversarial-model https://rfc.vac.dev/spec/13/#security-consideration

There we can write main assumptions, e.g:
- Poseidon primitive used and what happens if it is broken
- Groth16 used and what happens if trusted setup is broken (e.g.), see https://github.com/vacp2p/research/issues/9

b-d1 replied:
Agreed, I will add this section and also mention which params are used for the canonical poseidon.

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+### Hash function
+
+Canonical [Poseidon hash implementation](https://eprint.iacr.org/2019/458.pdf) is used, 
+as implemented in the circomlib library, according to the Poseidon paper.
```
- Link to implementation in circomlib
- Mention that circomlib poseidon is an example / (e.g. https://github.com/kilic/rln/blob/master/src/poseidon.rs should work too)

More generally, we can clarify this in implementation section that the concepts can be implemented separately (e.g. https://github.com/kilic/rln/), but for interop reasons circom-based on is canonical one.

b-d1 replied:
Yes I will add this, basically the broader conclusion would be that different posseidon hash functions can be used, and we're using the canonical one for interop reasons as described.

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+### Circuit parameters
+
+**Public Inputs**
+- `x` - signal hash
```
define signal hash

b-d1 replied:
yes good point, will be defined.

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+
+### Membership implementation
+
+For a valid signal, user's `identity commitment` (more on identity commitments below) must be exists in identity membership tree.
```
```suggestion
For a valid signal, a user's `identity commitment` must exist in the identity membership tree. More on identity commitments below.
```

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+
+# Motivation
+
+Spamming is the use of messaging systems to send multiple unsolicited messages (spam) to large numbers of recipients for the purpose of commercial advertising, for the purpose of non-commercial proselytizing, for any prohibited purpose (especially the fraudulent purpose of phishing), or simply sending the same message over and over to the same user. 
```
Agree to focus more on rate limiting, and then mention private economic spam protection as an example application, along with others

b-d1 replied:
Do you think anything more should be added here or is this section good?

oskarth commented on content/docs/rfcs/32/README.md:
````diff
@@ -0,0 +1,457 @@
+
+```
+{
+    identity_secret: secret_component_arr,
````
What is `secret_component_arr`? New term that should be introduced

b-d1 replied:
Yes I will fix this, this is left over from the previous writeup.

oskarth commented on content/docs/rfcs/32/README.md:
````diff
@@ -0,0 +1,457 @@
+}
+```
+
+For registration, the user needs to submit their `identity_commitment` (along with any additional registration requirements) to the registry party. 
````
Do we want to call it "registry party" more officially? Would be good if this was consistent/more formally defined. E.g. if it is a single source of truth (merkle tree on chain / server).

This makes it easier to reason about when there different views of the same canonical data (e.g. in p2p setting where the merkle tree might be out of date, or a mismatch could be introduced in race condition or whatever)

b-d1 replied:
No I couldn't find a better term for this, feel free to suggest what would be a better term for this. Maybe just `registry` is enough, but open for suggestions.

oskarth replied:
`registry` seems fine to me :)

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+For generating a proof, 
+the users need to obtain the required parameters or compute them themselves, 
+depending on the application implementation and client libraries supported by the application.
+For example the users can store the memebrship merkle tree on their end and 
```
```suggestion
For example the users can store the membership merkle tree on their end and 
```

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+### Signal hash
+
+The signal hash is generated by hashing the raw signal (or content) using the `keccak256` hash function, 
+converting the hashed value to number and padding it:
```
What is number?

One thing to be careful of here is making assumptions on how things work in JS. I don't see anything immediately that'd be inconsistent, but it might come up later. Can deal with in follow up issues as it becomes relevant :)

b-d1 replied:
Yes this might be too technical, and we should make it more general/abstract.

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+
+### Epoch
+
+The epoch (or external nullifier) is also obtained by hasing a raw string value using `keccak256`:
```
Rewrite for clarity - if I understand correctly this could be a new section, something like:

### External nullifier

The external nullifier is hashed using `keccak256` in the same way as for `epoch` above.

oskarth replied:
```suggestion
The epoch (or external nullifier) is also obtained by hashing a raw string value using `keccak256`:
```

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+### Merkle proof
+
+The merkle proof should be obtained locally or from a trusted third party. 
+By using the incremental merkle tree algorithm, 
```
Reference to this algorithm (can be to code or paper or blog post or whatever seems most canonical/useful)

b-d1 replied:
I will do, I think I have it in references but will link it explicitly here.

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+ the user's `identity_commitment` can be generated from the secret and it can be used for removing the user from the membership merkle tree (zeroing out the leaf that contains the user's `identity_commitment`). 
+Additionally, depending on the application the `identity_secret_hash` can be used for taking the user's provided stake.
+
+# Auxiliary tooling
```
What about making this a subsection under something like "implementation notes"? E.g. see https://rfc.vac.dev/spec/10/#appendix-c-implementation-notes

Then we can mention:

- Circom-based RLN is recommended because of small differences and having interop - also link to this Circuit implementation
- However existing ones like https://github.com/kilic/rln work
- Any other notes that are useful for people who would be *implementing* or analyzing RLN, or *using* it

b-d1 replied:
Seems like a great idea

oskarth commented on content/docs/rfcs/32/README.md:
````diff
@@ -0,0 +1,457 @@
+```
+
+
+# Flow
````
Perhaps we can make a short list of steps to make the flow more clear? After basic introduction perhaps.

E.g.

1. Registration
2. ...

We can also do a sequence diagram https://rfc.vac.dev/spec/10/#overview-of-protocol-interaction but might be overkill for now, up to you

b-d1 replied:
I agree, sequence diagram will be helpful for sure.

oskarth commented on content/docs/rfcs/32/README.md:
````diff
@@ -0,0 +1,457 @@
+    }
+```
+
+The generated proof (`zk_proof`) is composed of the following fields:
````
Elaborate on what these fields stand for (or link to reference that explains)

b-d1 replied:
Will do

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+  pi_b: [ [Array<string>], [Array<string>], [Array<string>] ],
+  pi_c: Array<string>
+  protocol: string,
+  curve: string
```
Are we making assumptions on curve? If so we should add this to primitives where Poseidon constants etc are defined

b-d1 replied:
No assumptions, maybe this part should be removed as it is too detailed and probably does not add anything significant to the specification?

oskarth replied:
I think it is fine to mention, but we can refer to more canonical place that deals with this type of thing. Where is this output from exactly and what does e.g. a, b, c correspond to here? Is this a general way of representing a proof?

If it is, and we know what is expected for each field, It means we can more easily compare different implementations, e.g. Rust based on etc to make sure proofs are identical.

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,457 @@
+
+
+
+### Sending proof
```
This head makes it seem like it is just sending proof, but it is sending "output message", which includes proof

b-d1 replied:
Yes, will be changed.

oskarth commented on content/docs/rfcs/32/README.md:
````diff
@@ -0,0 +1,457 @@
+The epoch (or external nullifier) is also obtained by hasing a raw string value using `keccak256`:
+
+```
+    raw_epoch: string = "1234"
````
There are quite a lot of fields and stuff going on, so I like using concrete examples. It'd be useful to have it for more messages, e.g. concrete values for epoch, limit etc.

Perhaps we can have an "example" section? E.g. like https://rfc.vac.dev/spec/16/#example-usage Either appended to each subsection, or as an "Example usage" section at top level

b-d1 replied:
Example usage section would be great. I will add it.

oskarth reviewed (commented):
Congrats on your first PR @bdim1! Looks very thorough, great job :)

I've added a bunch of comments to the PR with suggestions. I'll need a bit more time with some of the technical aspects to give better feedback here - I'm hoping to get there as https://github.com/vacp2p/zerokit gets built out.

Feel free to either incorporate suggestion this PR, or if it feels like a lot we can create a general "improvement issue/issues" for things we want to improve with spec later on.

oskarth commented:
Also @kilic might have some feedback on spec as he wrote https://github.com/kilic/rln

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: oskarth <ot@oskarthoren.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: oskarth <ot@oskarthoren.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: oskarth <ot@oskarthoren.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: oskarth <ot@oskarthoren.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: oskarth <ot@oskarthoren.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: oskarth <ot@oskarthoren.com>

b-d1 committed:
updates according to the pr notes

b-d1 committed:
Merge branch 'rfcs/rln-spec' of https://github.com/bdim1/rfc

 Conflicts:
	content/docs/rfcs/32/README.md

b-d1 commented:
@oskarth @staheri14 I've just pushed some updates reflecting the comments above. Haven't had the time to do a sequence diagram yet.

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+name: RLN specification
+status: raw
+editor: Blagoj Dimovski <blagoj.dimovski@yandex.com>
+contributors: <TO BE ADDED>
```
Field missing / should be blank if nothing

oskarth reviewed (commented):
CI is failing for some reason

> 9:22:29 AM: Error: Error building site: failed to render pages: render of "page" failed: "/opt/build/repo/layouts/docs/single.html:29:47": execute of template failed: template: docs/single.html:29:47: executing "main" at <.Params.contributors>: range can't iterate over <TO BE ADDED>

Seems to be related to contributors section not being filled out correctly

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+---
+slug: 32
+title: 32/RLN
+name: RLN specification
```
Rate Limit Nullifier maybe?

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+RLN (Rate limiting nullfier) is a construct based on zero-knowledge proofs that enables spam prevention mechanism for decentralized, anonymous environments. 
+In anonymous environments, the identity of the entities is unknown, and thus it is very hard to prevent spam attacks. 
+Many applications would benefit from anonymity, but the potential for spam could worsen their UX to a large extent. 
+RLN allows for anonymous signaling while enabling spam prevention on protocol level.
```
```suggestion
Rate limiting nullified (RLN) is a construct based on zero-knowledge proofs that provides an anonymous rate-limited signaling/messaging framework suitable for decentralized (and centralized) environments. 
Anonymity refers to the unlinkability of messages to their owner.
RLN guarantees a messaging rate is enforced cryptographically while preserving the anonymity of the message owners.
A wide range of applications can benefit from RLN and provide desirable security features. 
For example, an e-voting system can integrate RLN to contain the voting rate while protecting the voters-vote unlinkability. 
Another use case is to protect an anonymous messaging system against DDoS and spam attacks by containing messaging rate of users. 
This latter use case is explained in [17/Waku-RLN-Relay RFC](/spec/17).
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+RLN allows for anonymous signaling while enabling spam prevention on protocol level.
+
+
+# Technical overview
```
I think it would be easier to understand the specs if we switch the order of the Technical overview section with the Flow section.

b-d1 replied:
I agree with this.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+the users can generate ZK-Proofs for their signals and can participate in the application. 
+If a user generates more signals than allowed in an epoch, 
+the user risks being slashed - by revealing his secret and revealing his identity commitment. 
+If financial stake is put in place, the user also risks his stake being taken.
```
```suggestion
The users participate in the protocol by first registering to an application-defined group referred by the _membership group_.
Registration to the group is mandatory for signaling in the application. 
After registration, group members can generate Zero-knowledge Proof of membership for their signals and can participate in the application. 
Usually, the membership requires a financial or social stake which 
is beneficial for the prevention of Sybil attacks and double-signaling.
Group members are allowed to send one signal per epoch.
If a user generates more signals than allowed in an epoch, 
the user risks being slashed - by revealing his membership secret credentials. secret and revealing his identity commitment. 
If the financial stake is put in place, the user also risks his stake being taken.
```

staheri14 reviewed (commented):
Thanks Blagoj for preparing the RLN specs! Nice one! I put a few comments on the motivation part and the flow section. I tried to revise the text to present RLN as a rate-limiting construct rather than a spam-protection tool. I also think it would be helpful if we switch the order of Flow and Techincal overview sections. The flow section helps understand the overall construct and how technical terms are related and where they locate in the big picture. Then, the Technical overview section sheds light on details and the low-level constructions. What do you think about this reorganization?

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+|---------------------------|-------------------------------------------------------------------------------------|
+| **X Share**               | Hash of the signal, public input in the RLN circuit.    |
+| **A0**               | The identity secret hash.    |
+| **A1**               | Poseidon hash of [a0, external_nullifier] (see about External nullifier below).    |
```
```suggestion
| **A1**               | Poseidon hash of [A0, External nullifier] (see about External nullifier below).    |
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+
+| Term           | Description                                                       |
+|---------------------------|-------------------------------------------------------------------------------------|
+| **X Share**               | Hash of the signal, public input in the RLN circuit.    |
```
How about (we also use `share_x` in the waku-rln-relay spec):
```suggestion
| **share_x**               | Hash of the signal.    |
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+| **X Share**               | Hash of the signal, public input in the RLN circuit.    |
+| **A0**               | The identity secret hash.    |
+| **A1**               | Poseidon hash of [a0, external_nullifier] (see about External nullifier below).    |
+| **Y Share**          | The result of the polynomial equation (y = a0 + a1*x). Public output of the circuit.    |
```
Similar to my suggestion for X share:
```suggestion
| **share_y**          | The result of the polynomial equation (y = a0 + a1*x). The public output of the circuit.    |
```

staheri14 replied:
Wondering Y share is the public output of the circuit? or public input?

b-d1 replied:
It is output from the circuit, but it is also input needed for proof verification.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+| Term                      | Description                                                                         |
+|---------------------------|-------------------------------------------------------------------------------------|
+| **ZK-SNARK**    | https://z.cash/technology/zksnarks/              |
+| **Stake**                 | Financial or social stake required for registering in the RLN applications. The staking requirement is necessary for disincentivising the end users to spam. A common stake examples are: locking cryptocurrency (financial), linking reputable social identity. |
```
I think this part can be removed safely to keep the table more succinct, it is already explained in the text 
```suggestion
| **Stake**                 | Financial or social stake required for registering in the RLN applications.  Common stake examples are: locking cryptocurrency (financial), linking reputable social identity. |
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+| **ZK-SNARK**    | https://z.cash/technology/zksnarks/              |
+| **Stake**                 | Financial or social stake required for registering in the RLN applications. The staking requirement is necessary for disincentivising the end users to spam. A common stake examples are: locking cryptocurrency (financial), linking reputable social identity. |
+| **RLN Registration**    | The process of the user registering themselves to the RLN protocol. The registrations can occur onchain or offchain. Financial or social stake is required for registering to the protocol.   |
+| **Slashing**              | The process of retreiving the secret hash from the user that spams, and from the secret hash generating the identity commitment. Having the secret hash of a user implies that anyone could reveal their "identity" and also retreive their financial stake if it is enabled by the app.      |
```
I am not sure whether we need these two i.e., RLN registration and slashing as terms

b-d1 replied:
Probably not, I will remove them.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+| **Slashing**              | The process of retreiving the secret hash from the user that spams, and from the secret hash generating the identity commitment. Having the secret hash of a user implies that anyone could reveal their "identity" and also retreive their financial stake if it is enabled by the app.      |
+| **Identity nullifier**    | Random 32 byte value used as component for identity secret generation.              |
+| **Identity trapdoor**     | Random 32 byte value used as component for identity secret generation.              |
+| **Identity secret**       | An array of two unique random components (identity nullifier and identity trapdoor), which must be kept secret from the users. Secret hash and identity commitment are derived from this array.        |
```
Reordered identity secret row:
```suggestion
| **Identity secret**       | An array of two unique random components (identity nullifier and identity trapdoor), which must be kept secret from the users. Secret hash and identity commitment are derived from this array.        |
| **Identity nullifier**    | Random 32 byte value used as component for identity secret generation.              |
| **Identity trapdoor**     | Random 32 byte value used as component for identity secret generation.              |
```

staheri14 replied:
```suggestion
| **Identity nullifier**    | Random 32 byte value used as component for identity secret generation.              |
| **Identity trapdoor**     | Random 32 byte value used as component for identity secret generation.              |
| **Identity secret**       | An array of two unique random components (identity nullifier and identity trapdoor), which must be kept secret.        |
```

staheri14 reviewed (commented):
A few more editorial comments.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+| **Identity nullifier**    | Random 32 byte value used as component for identity secret generation.              |
+| **Identity trapdoor**     | Random 32 byte value used as component for identity secret generation.              |
+| **Identity secret**       | An array of two unique random components (identity nullifier and identity trapdoor), which must be kept secret from the users. Secret hash and identity commitment are derived from this array.        |
+| **Identity secret hash**  | The hash of the identity secret, obtained using the Poseidon hash function. It is used for deriving the identity commitment of the user, and as a private input for zk proof generation. The secret hash should be treated as "private key" for the RLN protocol. The anonymity and security of the user is dependent on the secret hash.                                       |
```
Can you please on this part? the "private key" used in the first sentence implies some level of secrecy for the secret hash but the next sentence 
> "The anonymity and security of the user is dependent ..." 

defeats this implication. Wondering should the secret hash be kept private or not?

b-d1 replied:
it should be definitely kept private, I should formulate this better as this sentence is confusing

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+| **Identity secret**       | An array of two unique random components (identity nullifier and identity trapdoor), which must be kept secret from the users. Secret hash and identity commitment are derived from this array.        |
+| **Identity secret hash**  | The hash of the identity secret, obtained using the Poseidon hash function. It is used for deriving the identity commitment of the user, and as a private input for zk proof generation. The secret hash should be treated as "private key" for the RLN protocol. The anonymity and security of the user is dependent on the secret hash.                                       |
+| **Identity commitment**   | Hash obtained from the `Identity secret hash` by using the poseidon hash function. It is used by the users for registering in the protocol.                                       |
+| **Signal**                | The message that needs to be sent by the user in a privacy preserving way. The message can be of different type, depending on the application. It can be a chat message, url request, protobuf message. |
```
```suggestion
| **Signal**                | The message generated by a user. It is an arbitrary bit string that may represent a chat message, a URL request, protobuf message, etc. |
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+| **Identity commitment**   | Hash obtained from the `Identity secret hash` by using the poseidon hash function. It is used by the users for registering in the protocol.                                       |
+| **Signal**                | The message that needs to be sent by the user in a privacy preserving way. The message can be of different type, depending on the application. It can be a chat message, url request, protobuf message. |
+| **Signal hash**           | Hash of the signal, used as an input in the RLN circuit. |
+| **Epoch**                 | An identifier that groups signals and determines the validity of the signal. Can be though ot as voting booth. Usually a timestamp or time interval in the RLN apps. |
```
```suggestion
| **Epoch**                 | An identifier that groups signals and can be thought of as a voting booth. Usually a timestamp or time interval in the RLN apps. |
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+| **A0**               | The identity secret hash.    |
+| **A1**               | Poseidon hash of [a0, external_nullifier] (see about External nullifier below).    |
+| **Y Share**          | The result of the polynomial equation (y = a0 + a1*x). Public output of the circuit.    |
+| **External nullifier**    | `keccak256` hash of the Epoch.    |
```
Is it necessary to hash the Epoch? is it for length preserving? what if the epoch length is already less than 256 bits and the application uses zero padding to meet the length requirement?

b-d1 replied:
It is necessary. The necessity is because of the fact we need to have a deterministic format, so that we can ensure that hashing inside the circuit is done without any errors. Not parsing the epoch could lead to problems in the circuit, so internal representation of the epoch is needed. Of course other forms of internal representation could be used.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+| **A1**               | Poseidon hash of [a0, external_nullifier] (see about External nullifier below).    |
+| **Y Share**          | The result of the polynomial equation (y = a0 + a1*x). Public output of the circuit.    |
+| **External nullifier**    | `keccak256` hash of the Epoch.    |
+| **Internal nullifier**    | Poseidon hash of [a1, rln_identifier]. This field ensures that a user can send only one valid signal per epoch without risking being slashed. Public output of the circuit.   |
```
```suggestion
| **Internal nullifier**    | Poseidon hash of [A1, RLN Identifier]. This field ensures that a user can send only one valid signal per epoch without risking being slashed. Public input of the circuit.   |
```

staheri14 reviewed (commented):
Added a few more comments.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+
+## ZK Circuits specification
+
+Anonymous signaling with spam detection is enabled by proving that the user is part of a group which has high barriers to entry (form of stake) and 
```
```suggestion
Anonymous signaling with a controlled rate limit is enabled by proving that the user is part of a group which has high barriers to entry (form of stake) and 
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,565 @@
+## ZK Circuits specification
+
+Anonymous signaling with spam detection is enabled by proving that the user is part of a group which has high barriers to entry (form of stake) and 
+enabling secret reveal if more signals are produced than the spam threshold (system parameter) per epoch.
```
I'd suggest focusing on the general application of RLN, and replacing the term `spam`  with `double signaling`, also using `rate limit` or `double signaling detection` instead of `spam detection`.

b-d1 replied:
I agree with your suggestion, it's much more precise.

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
fix: reformatting

b-d1 commented:
@oskarth @staheri14 I've just updated the RFC spec by applying the suggestions/comments up to this point.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+
+# Motivation
+
+Rate limiting nullified (RLN) is a construct based on zero-knowledge proofs that provides an anonymous rate-limited signaling/messaging framework suitable for decentralized (and centralized) environments. 
```
```suggestion
Rate limiting nullifier (RLN) is a construct based on zero-knowledge proofs that provides an anonymous rate-limited signaling/messaging framework suitable for decentralized (and centralized) environments. 
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+is beneficial for the prevention of Sybil attacks and double-signaling.
+Group members are allowed to send one signal per epoch.
+If a user generates more signals than allowed in an epoch, 
+the user risks being slashed - by revealing his membership secret credentials. secret and revealing his identity commitment. 
```
```suggestion
the user risks being slashed - by revealing his membership secret credentials. 
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+After registration, group members can generate Zero-knowledge Proof of membership for their signals and can participate in the application. 
+Usually, the membership requires a financial or social stake which 
+is beneficial for the prevention of Sybil attacks and double-signaling.
+Group members are allowed to send one signal per epoch.
```
Since the spec is not gonna be focused on spam protection use case, then I think the `epoch` should be replaced with `external nullifier` throughout the spec.

staheri14 replied:
Also,  `epoch` (or external nullifier) is not conceptually defined at this point.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+- centralized registrations, by using a central server
+- decentralized registrations, by using a smart contract
+
+What is important is that the user's identity commitment is stored in a merkle tree, 
```
```suggestion
What is important is that the users' identity commitments (explained in section [User Indetity](#user-identity)) are stored in a Merkle tree, 
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+- decentralized registrations, by using a smart contract
+
+What is important is that the user's identity commitment is stored in a merkle tree, 
+and the users can obtain a merkle proof proving that they are part of the group.
```
```suggestion
and the users can obtain a Merkle proof proving that they are part of the group.
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+usually a financial or social stake is introduced.
+
+An example for financial stake is: For each registration a certain amount of ETH is required.
+An example for social stake is using InterRep as a registry - 
```
A ref to Unirep would be great.

staheri14 commented on content/docs/rfcs/32/README.md:
````diff
@@ -0,0 +1,560 @@
+```
+
+For registration, the user needs to submit their `identity_commitment` (along with any additional registration requirements) to the registry. 
+Upon registration, they should receive `leaf_index` value which represents their position in the merkle tree. 
````
```suggestion
Upon registration, they should receive a `leaf_index` value which represents their position in the Merkle tree. 
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+After registration, 
+the users can participate in the application by sending signals to the other participants in a decentralised manner or to a centralised server. 
+Along with their signal, 
+they need to generate a ZK-Proof by using the circuit with the specification described above. 
```
there is no circuit specification `above`, you may put a link to the corresponding section.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+For generating a proof, 
+the users need to obtain the required parameters or compute them themselves, 
+depending on the application implementation and client libraries supported by the application.
+For example the users can store the membership merkle tree on their end and 
```
```suggestion
For example, the users can store the membership Merkle tree on their end and 
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+the users need to obtain the required parameters or compute them themselves, 
+depending on the application implementation and client libraries supported by the application.
+For example the users can store the membership merkle tree on their end and 
+generate a merkle proof whenever they want to generate a signal. 
```
```suggestion
generate a Merkle proof whenever they want to generate a signal. 
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+they need to generate a ZK-Proof by using the circuit with the specification described above. 
+
+For generating a proof, 
+the users need to obtain the required parameters or compute them themselves, 
```
```suggestion
the users need to have access to the recent state of the membership Merkle tree.
They may obtain this information by querying from others or by computing locally, 
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+generate a merkle proof whenever they want to generate a signal. 
+
+### Implementation notes
+
```
```suggestion
The information required for proof generation is as follows:
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+
+The signal hash can be generated by hashing the raw signal (or content) using the `keccak256` hash function.
+
+#### Epoch
```
```suggestion
#### External Nullifier
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+
+#### Epoch
+
+The epoch can be generated by hashing a raw string (i.e UNIX timestamp) value using `keccak256`.
```
This definition is not general (it is specific to the spam protection application).

staheri14 commented on content/docs/rfcs/32/README.md:
````diff
@@ -0,0 +1,560 @@
+```
+{
+  root: bigint
+  indices: number[]
````
Can you please add an explanation for each of these fields?

staheri14 replied:
You may have missed this comment 👉  https://github.com/vacp2p/rfc/pull/482/files#r814354721

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+      identity_path_index: merkle_proof.indices,
+      x: signal_hash,
+      epoch: epoch,
+      rln_identifier: rln_identifier
```
The rln identifier is not introduce yet in the spec

staheri14 commented on content/docs/rfcs/32/README.md:
````diff
@@ -0,0 +1,560 @@
+
+```
+    {
+      identity_secret: identity_secret_hash,
````
I think there is no need to indicate these fields as key-value pairs, is there any specific reason for this representation?

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+
+The proof output is calculated locally, 
+in order for the required fields for proof verification to be sent along with the proof. 
+The proof output is composed of the `y` share of the secret equation and the `internal_nullifier`. 
```
The secret sharing is not yet explained

staheri14 commented on content/docs/rfcs/32/README.md:
````diff
@@ -0,0 +1,560 @@
+The following fields are needed for proof output calculation:
+
+```
+{
````
Isn't this the same as the parameters explained in line 125? We may be able to remove this one if they are duplicates

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+  identity_secret_hash: bigint, 
+  epoch: bigint, 
+  rln_identifier: bigint,
+  x: bigint, 
```
btw, why path elements are excluded?  isn't it needed for proof output calculation?

staheri14 commented on content/docs/rfcs/32/README.md:
````diff
@@ -0,0 +1,560 @@
+}
+```
+
+The output `[y, internal_nullifier]` is calculated in the following way:
````
I'd suggest adding the description of Shamir secret sharing (or just the linear secret sharing) in here and incorporating the a_0 and a_1 equations in that description.

b-d1 replied:
this has been address already

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+
+y = a_0 + x * a_1
+
+internal_nullifier = poseidonHash([a_1, rln_identifier])
```
would be good to also add a bride description of the role of internal nullified (like a fingerprint for users in the given external nullifier)

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+    content: signal, # non-hashed signal
+    proof: zk_proof,
+    internal_nullifier: internal_nullifier,
+    x: x, # signal_hash
```
I think later in the spec we use the term share_x and share_y for x and y, it is better to use a consistent notation.

staheri14 replied:
@bdim1 I thought you may have missed this comment ☝️

b-d1 replied:
sorry @staheri14 but everywhere now x and y are used. Except the storage container metadata variable namings, but the notation there is explicit enough to account for these differences (x_shares and y_shares). Also the context where these are used is very clear:

```
If the `x` and `y` fields exist in the `x_shares` and `y_shares` array for the `external_nullifier` and the `internal_nullifier` the message can be considered as a duplicate. 
Duplicate messages are discarded.
```

staheri14 commented on content/docs/rfcs/32/README.md:
````diff
@@ -0,0 +1,560 @@
+
+```
+{
+    content: signal, # non-hashed signal
````
wondering why the key-value pair format? 
e.g., `signal` is already a named variable, it is a bit confusing to use the `content` term for it.  Can `content` contain anything rather than `signal`?

b-d1 replied:
You're right, let's just name it signal

staheri14 commented on content/docs/rfcs/32/README.md:
````diff
@@ -0,0 +1,560 @@
+```
+  {
+      root: merkle_proof.root,
+      epoch: epoch
````
Why they are not defaults?  I think it would be prudent to include them in the required fields (as default behavior), but then saying they may be excluded if the application wants so. I think typically epoch and root are needed to verify the proof unless they are somehow available at the verifier side.

b-d1 replied:
They are not default because they are implementation specific. Depending on the app, they can only be used at the verifier side, and not actually be included in the messages themselves.

staheri14 reviewed (commented):
Thanks Blagoj for revising the PR and addressing the comments! 
I've put a few more comments and questions.

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+---
+slug: 32
+title: 32/RLN
+name: Rate Limit Nullifier specification
```
```suggestion
name: Rate Limit Nullifier
```

oskarth commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,560 @@
+
+# Abstract
+
+The following document covers the specification of the RLN construct as well as the auxiliary libraries necessary for interaction with it.
```
```suggestion
The following specification covers the RLN construct as well as some auxiliary libraries useful for interacting with it.
```

oskarth replied:
> Rate limiting nullified (RLN) is a construct based on zero-knowledge proofs that provides an anonymous rate-limited signaling/messaging framework suitable for decentralized (and centralized) environments. 

I wonder if this doesn't belong in abstract aka "what is this", then motivation section is more "why do we need this"

oskarth reviewed (approved):
Awesome work! While there are probably places we can tweak and improve, this is definitely good enough for a raw spec :)

What do you see as necessary for it to become a draft spec? Do we want to make sure the nim-waku version matches the js-waku version? I.e. using Circom based representation

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: oskarth <ot@oskarthoren.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: oskarth <ot@oskarthoren.com>

b-d1 committed:
updates

b-d1 commented:
The changes from the comments above have been applied. @oskarth @staheri14

b-d1 commented:
> Awesome work! While there are probably places we can tweak and improve, this is definitely good enough for a raw spec :)
> 
> What do you see as necessary for it to become a draft spec? Do we want to make sure the nim-waku version matches the js-waku version? I.e. using Circom based representation

not sure about this, have to check what are the requirements and prerequisites for the spec become a draft spec

staheri14 commented:
Re this https://github.com/vacp2p/rfc/pull/482#discussion_r812868232

@bdim1 
Can you please give more details about the desired format of ext nullifier? what kind of error can happen if an arbitrary string of 256 bits is sent (without being hashed)?

staheri14 commented:
Re this https://github.com/vacp2p/rfc/pull/482#discussion_r814708206

> They are not default because they are implementation specific. Depending on the app, they can only be used at the verifier side, and not actually be included in the messages themselves.

@bdim1

Right (it can be optional), however, these represent the root and epoch based on which the proof of the message is generated (so these are public inputs for the proof verification). Of course, a verifier may also have access to this info, but if we exclude them (in general) there would be insufficient data for the proof verification.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+Each user of the protocol (server or otherwise) will need to store metadata for each message received by each user, 
+for the given external_nullifier. 
+The data can be deleted when the external_nullifier passes. 
+Storing metadata is required, so that if a user sends more than one unique signal per external_nullifier, 
```
```suggestion
for the given `external_nullifier`. 
The data can be deleted when the `external_nullifier` passes. 
Storing metadata is required, so that if a user sends more than one unique signal per `external_nullifier`, 
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+The metadata stored contains the `x`, `y` shares and the `internal_nullifier` for the user for each message. 
+If enough such shares are present, the user's secret can be retreived.
+
+One way of storing received metadata (`messaging_metadata`) is the following format: 
```
Is it a JSON format?

b-d1 replied:
Yes it is pseudo json format, where `[external_nullifier]` and `[internal_nullifier]` represent that values for the respective external_nullifier and internal_nullifier should be provided.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+**1. `external_nullifier` correctness**
+Upon received `output_message`, first the `external_nullifierexternal_nullifier` field is checked, 
+to ensure that the message matches the current external_nullifier. 
+If the external_nullifier is correct the verification continues, otherwise the message is discarded.
```
```suggestion
Upon received `output_message`, first the `external_nullifier` field is checked, 
to ensure that the message matches the current `external_nullifier`. 
If the `external_nullifier` is correct the verification continues, otherwise, the message is discarded.
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+If the proof verification is correct, 
+the verification continues, otherwise the message is discarded.
+
+**4. spam verification**
```
```suggestion
**4. Double signaling verification**
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+
+**4. spam verification**
+
+After the proof is verified The `x`, and `y` fields are added to the `x_shares` and `y_shares` arrays of the `messaging_metadata` `external_nullifier` and `internal_nullifier` object. 
```
```suggestion
After the proof is verified the `x`, and `y` fields are added to the `x_shares` and `y_shares` arrays of the `messaging_metadata` `external_nullifier` and `internal_nullifier` object. 
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+**4. spam verification**
+
+After the proof is verified The `x`, and `y` fields are added to the `x_shares` and `y_shares` arrays of the `messaging_metadata` `external_nullifier` and `internal_nullifier` object. 
+If the length of the arrays is equal to the spam threshold (`limit`), the user can be slashed.
```
Isn't fixed to 2?

```suggestion
If the length of the arrays is equal to the signaling threshold (`limit`), the user can be slashed.
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+
+#### Slashing
+
+After the verification, the user can be slashed if enough shares are present to reconstruct their `identity_secret_hash` from `x_shares` and `y_shares` fields, 
```
`identity_secret_hash`  or `identity_secret`?

 ```suggestion
After the verification, the user can be slashed if enough shares are present to reconstruct their `identity_secret` from `x_shares` and `y_shares` fields, 
```

b-d1 replied:
Identity secret hash, identity secret is an array of two elements: [identity_nullifier, identity_trapdoor]

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+After the verification, the user can be slashed if enough shares are present to reconstruct their `identity_secret_hash` from `x_shares` and `y_shares` fields, 
+for their `internal_nullifier`.
+The secret can be retreived by the properties of the Shamir's secret sharing scheme. 
+In particular the secret (`a_0`) can be retrieved by computing `Lagrange polynomials`. 
```
A ref to the Lagrange interpolation would be good

b-d1 replied:
Yes, will add that.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+|---------------------------|-------------------------------------------------------------------------------------|
+| **ZK-SNARK**    | https://z.cash/technology/zksnarks/              |
+| **Stake**                 | Financial or social stake required for registering in the RLN applications.  Common stake examples are: locking cryptocurrency (financial), linking reputable social identity. |
+| **Identity secret**       | An array of two unique random components (identity nullifier and identity trapdoor), which must be kept private by the user. Secret hash and identity commitment are derived from this array.        |
```
The mathematical relation between identity trapdoor and identity secret is not explained in the specs, maybe just add a todo to include its description later

b-d1 replied:
Yes, but that is out of the scope for this document. The nullifier and trapdoor namings are used here to preserve consistency with other protocols (where these fields are actually used more). Here we only use them as a basis to derive a hash from them.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+| **Identity secret**       | An array of two unique random components (identity nullifier and identity trapdoor), which must be kept private by the user. Secret hash and identity commitment are derived from this array.        |
+| **Identity nullifier**    | Random 32 byte value used as component for identity secret generation.              |
+| **Identity trapdoor**     | Random 32 byte value used as component for identity secret generation.              |
+| **Identity secret hash**  | The hash of the identity secret, obtained using the Poseidon hash function. It is used for deriving the identity commitment of the user, and as a private input for zk proof generation. The secret hash should be kept private by the user.                            |
+| **Identity commitment**   | Hash obtained from the `Identity secret hash` by using the poseidon hash function. It is used by the users for registering in the protocol.                                       |
```
Why two layers of hashing are needed? `id comm = h(h(secret id key))` why not `id comm = h(secret id key)`?

b-d1 replied:
This is a great question. The hashing scheme used here is based on the design decisions which also include the Semaphore circuits. Our goal was to ensure compatibility of the secrets for apps that use Semaphore and RLN circuits while also not compromising on security because of using the same secrets.

For example let's say there is a voting app that uses Semaphore, and also a chat app that uses RLN. The UX would be better if the users would not need to care about complicated identity management (secrets and commitments) they use for each app, and it would be much better if they could use a single id commitment for this. Also in some cases these kind of dependency is required - RLN chat app using Interep as a registry (instead of using financial stake). 
One potential concern about this interoperability is a slashed user on the RLN app side having their security compromised on the semaphore side apps as well. I.e obtaining the user's secret, anyone would be able to generate valid semaphore proofs as the slashed user. We don't want that, and we should keep user's app specific security threats in the domain of that app alone.

To achieve the above interoperability UX while preventing the shared app security model (i.e slashing user on an RLN app having impact on Semaphore apps), we had to do the follow in regard the identity secret and identity commitment:

Identity secret = [nullifier, trapdoor]
Identity secret hash = poseidon(Identity secret)
Identity commitment = poseidon(Identity secret hash)

Secret components for generating Semaphore proof:
- nullifier
- trapdoor

Secret components for generting RLN proof:
- identity secret hash

When a user is slashed on the RLN app side, their identity secret hash is revealed. However a semaphore proof can't be generated because we do not know the user's nullifier and trapdoor.
With this design we achieve:

id commitment (Semaphore) == id commitment (RLN)
secret (semaphore) != secret (RLN).

This is the only option we had for the scheme in order to satisfy the properties described above.

Also for RLN we do a single secret component input for the circuit. Thus we need to hash the secret array (two components) to a secret hash, and we use that as a secret component input.

staheri14 replied:
Thanks for the very clear explanation.  This is an important design choice and is worth being included at the end of the spec in a separate section like `#Security consideration`  or `#compatibility with Semaphore`.

b-d1 replied:
Added in a new appendix:
Appendix B: Identity scheme choice

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+| **Identity secret hash**  | The hash of the identity secret, obtained using the Poseidon hash function. It is used for deriving the identity commitment of the user, and as a private input for zk proof generation. The secret hash should be kept private by the user.                            |
+| **Identity commitment**   | Hash obtained from the `Identity secret hash` by using the poseidon hash function. It is used by the users for registering in the protocol.                                       |
+| **Signal**                | The message generated by a user. It is an arbitrary bit string that may represent a chat message, a URL request, protobuf message, etc. |
+| **Signal hash**           | Hash of the signal, used as an input in the RLN circuit. |
```
Poseidon hash (not Keccak)?

b-d1 replied:
Signal hash is keccak hash, will update this.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+| **Identity secret**       | An array of two unique random components (identity nullifier and identity trapdoor), which must be kept private by the user. Secret hash and identity commitment are derived from this array.        |
+| **Identity nullifier**    | Random 32 byte value used as component for identity secret generation.              |
+| **Identity trapdoor**     | Random 32 byte value used as component for identity secret generation.              |
+| **Identity secret hash**  | The hash of the identity secret, obtained using the Poseidon hash function. It is used for deriving the identity commitment of the user, and as a private input for zk proof generation. The secret hash should be kept private by the user.                            |
+| **Identity commitment**   | Hash obtained from the `Identity secret hash` by using the poseidon hash function. It is used by the users for registering in the protocol.                                       |
```
All the hashed in these two lines refer to Poseidon hash?

b-d1 replied:
Identity secret hash and Identity commitment are obtained using the Poseidon hash function.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+
+| Term           | Description                                                       |
+|---------------------------|-------------------------------------------------------------------------------------|
+| **x**               | Hash of the signal.    |
```
Poseidon hash?

b-d1 replied:
No, Keccak hash

staheri14 commented on content/docs/rfcs/32/README.md:
````diff
@@ -0,0 +1,554 @@
+
+```
+a_0 = identity_secret
+a_1 = poseidonHash([a0, external_nullifier])
````
I thought the idea of having an rln identifier was to make the secret-sharing application-specific so that users can use the same rln group for different applications, however, in this formulation, rln identifier is not used at all. It is only used in the internal nullifier. Shouldn't the rln identifier be embedded in the `a1` coefficient?

b-d1 replied:
The role of RLN identifier is to obfuscate the nullifer so that a user generating a signal for two different apps at the same epoch is impossible to identify (impossible without additional extra metadata, the additional metadata being the user's secret).  For this reason the RLN identifier is used in the internal nullifier as a strong enough measure to prevent this issues. There are potentially other solutions for this, I am not 100% sure if this can be embedded in the a1 coefficient without having significant impact on the other parts of the architecture.

staheri14 replied:
> that a user generating a signal for two different apps at the same epoch is impossible to identify

One way is to do a brute-force on the published secret shares (of all the peers) for each epoch to recover various secret keys and see if their commitment appears in the membership group.

b-d1 replied:
Agreed, the brute force attacks possibilities are open, especially if:
- the same epoch scheme is used across different applications
- an epoch is chosen inappropriately ("larger" epoch increases the probability of messages being sent to different application in the same epoch) 
- the anonymity sets are small

We'll look further on how to resolve this problem, and also reconsider the possibility of brute force attacks.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+
+Canonical [Poseidon hash implementation](https://eprint.iacr.org/2019/458.pdf) is used, 
+as implemented in the [circomlib library](https://github.com/iden3/circomlib/blob/master/circuits/poseidon.circom), according to the Poseidon paper.
+This Poseidon hash version (canonical implementation) is uses the following parameters:
```
```suggestion
This Poseidon hash version (canonical implementation) uses the following parameters:
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+in order for their proof to be verified. 
+`x` is the hashed signal, while the `y` is the circuit output. 
+With more than one pair of unique shares, anyone can derive `a_0`, the `identity_secret_hash` . 
+Hash of a signal will be evaluation point x. 
```
```suggestion
The hash of a signal will be evaluation point `x`. 
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+`x` is the hashed signal, while the `y` is the circuit output. 
+With more than one pair of unique shares, anyone can derive `a_0`, the `identity_secret_hash` . 
+Hash of a signal will be evaluation point x. 
+So that a member who sends more that one unique signal per external_nullifier risks their identity secret being revealed.
```
```suggestion
So that a member who sends more than one unique signal per `external_nullifier` risks their identity secret being revealed.
```

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+Hash of a signal will be evaluation point x. 
+So that a member who sends more that one unique signal per external_nullifier risks their identity secret being revealed.
+
+Note that shares used for different external nullifiers and different RLN apps cannot be used to derive the secret key.
```
Related to this, please see my comment on line 156.

b-d1 replied:
Yes, addressed the nullifier question there. I will update the text to be more clear regarding the nullifier.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+the user uses the same credentials and sends a different message for two different RLN apps using the same `external_nullifier`,
+then their secret key can be revealed. 
+With adding the `rln_identifier` field we obscure the nullifier, 
+so this kind of attack cannot happen. 
```
Obscuring the internal nullifier does not alone make the attack impossible, an attacker with some side knowledge can still target a particular peer knowing that she uses the same rln credentials for two different apps.

b-d1 replied:
This is true, but that side knowledge would be the internal nullifier itself. Otherwise it is very hard to keep track of whose messages are the targeted user's one. Think of the internal nullifier as a user's epoch based identity. If that is different per app, you can't really find out which messages are which. However if there is an entity that can link messages to the specific user (i.e they identify the user by a certain identifiable property), then this would be problematic but then we have larger problems than the nullifier itself.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+RLN is an experimental and still un-audited technology. This means that the circuits have not been yet audited. 
+Another consideration is the security of the underlying primitives. 
+zk-SNARKS require a trusted setup for generating a prover and verifier keys.
+The standard for this is to use trusted [Multi-Party Computation (MPC)](https://en.wikipedia.org/wiki/Secure_multi-party_computation) ceremony, 
```
I'd suggest linking the original MPC for groth16 (in replace of the Wikipedia link)

b-d1 replied:
MPC for the RLN circuits has not yet been done, would be happy to replace the wikipedia links if you think there is something more appropriate that describes the concept.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,554 @@
+zk-SNARKS require a trusted setup for generating a prover and verifier keys.
+The standard for this is to use trusted [Multi-Party Computation (MPC)](https://en.wikipedia.org/wiki/Secure_multi-party_computation) ceremony, 
+which requires two phases. 
+Trusted MPC ceremony has not yet been performed for the RLN circuits.
```
I think it is done, I will find a link and share it with you.

b-d1 replied:
MPC for the RLN circuits has not yet been done.

staheri14 replied:
Right, the one I knew about was for Semaphore

staheri14 reviewed (commented):
Thanks for addressing the comments!
I have done my final round of reviews. After this set of comments, from my pov, the RFC looks good for the raw version and is ready to go live. 
Congratulations on your first Vac RFC PR in advance!

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

b-d1 committed:
content updates

b-d1 commented:
> Re this [#482 (comment)](https://github.com/vacp2p/rfc/pull/482#discussion_r812868232)
> 
> @bdim1 Can you please give more details about the desired format of ext nullifier? what kind of error can happen if an arbitrary string of 256 bits is sent (without being hashed)?

Yes. The desired format is a poseidon hashable value (the external_nullifier is used inside the circuit for hashing operation). I am not sure in what would happen exactly with an arbitrary string (this again depends on what is in the string), what can go wrong is the value being unable to be hashed, thus the circuit would throw an error.

b-d1 commented:
> Re this [#482 (comment)](https://github.com/vacp2p/rfc/pull/482#discussion_r814708206)
> 
> > They are not default because they are implementation specific. Depending on the app, they can only be used at the verifier side, and not actually be included in the messages themselves.
> 
> @bdim1
> 
> Right (it can be optional), however, these represent the root and epoch based on which the proof of the message is generated (so these are public inputs for the proof verification). Of course, a verifier may also have access to this info, but if we exclude them (in general) there would be insufficient data for the proof verification.

Yes you are definitely right. The fields are optional on message level, but definitely required on a proof verification level. They are optional on message level because these fields might be  used for message verification. Having the  root supplied at verification side can ensure that the proof is generated with a valid root, and not with an invalid root for example.

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,556 @@
+name: Rate Limit Nullifier
+status: raw
+editor: Blagoj Dimovski <blagoj.dimovski@yandex.com>
+contributors: 
```
I'd like to suggest the following contributors
Barry (don't know his contact), Sanaz Taheri <sanaz@status.im>, and Oskar Thorén <oskar@status.im>

b-d1 replied:
Added, will ask Barry for contact info and will add that.

b-d1 replied:
Barry's contact info updated

staheri14 reviewed (commented):
Thanks Blagoj, just some final comments, and we are almost done.

b-d1 committed:
rln updates

b-d1 commented:
@staheri14 I've just updated the spec, that address the latest comments above.

b-d1 committed:
feat: contributor address addition

staheri14 commented on content/docs/rfcs/32/README.md:
```diff
@@ -0,0 +1,614 @@
+name: Rate Limit Nullifier
+status: raw
+editor: Blagoj Dimovski <blagoj.dimovski@yandex.com>
+contributors: Barry Whitehat <barrywhitehat@protonmail.com>, Sanaz Taheri <sanaz@status.im>, Oskar Thorén <oskar@status.im>
```
```suggestion
contributors: 
- Barry Whitehat <barrywhitehat@protonmail.com>
- Sanaz Taheri <sanaz@status.im>
- Oskar Thorén <oskar@status.im>
```
The CI errors, ad far as I can tell, are related to the format of contributors, they should be itemized. My suggested change should fix the error.

staheri14 reviewed (approved):
Thanks Blagoj, I think it is ready to be merged, just two points before final merge
1- The CI checks are failing and I think the reason is as I explained in my [comment](https://github.com/vacp2p/rfc/pull/482#discussion_r826458425) 
2- Please use the _squash and merge_ option to merge the PR 

Thanks! excited to see that finally there is going to be a reference RFC for RLN!

oskarth committed:
Update content/docs/rfcs/32/README.md

Co-authored-by: Sanaz Taheri Boshrooyeh <35961250+staheri14@users.noreply.github.com>

oskarth commented:
Wohoo!

> 1- The CI checks are failing and I think the reason is as I explained in my https://github.com/vacp2p/rfc/pull/482#discussion_r826458425

Updated this and CI is now happy

@bdim1 currently we have it setup so all commits are signed. Would you mind rebasing/squashing so they are signed? Alternatively if you prefer I can also override and merge without code signing.

b-d1 committed:
Merge branch 'vacp2p:master' into rfcs/rln-spec

oskarth commented:
PR superseded by https://github.com/vacp2p/rfc/pull/498, closing this one
